src/samples/audio/smart_amp_test.c

```diff
@@ -60,22 +60,30 @@ static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
 
 	/* component model data handler */
 	sad->model_handler = comp_data_blob_handler_new(dev);
+	if (!sad->model_handler)
+		goto fail;
 
 	cfg = (struct sof_smart_amp_config *)ipc_sa->data;
 	bs = ipc_sa->size;
 
 	if ((bs > 0) && (bs < sizeof(struct sof_smart_amp_config))) {
 		comp_err(dev, "smart_amp_new(): failed to apply config");
-
-		rfree(sad);
-		return NULL;
+		goto fail;
 	}
 
 	memcpy_s(&sad->config, sizeof(struct sof_smart_amp_config), cfg, bs);
 
 	dev->state = COMP_STATE_READY;
 
 	return dev;
+
+fail:
+	if (sad) {
+		comp_data_blob_handler_free(sad->model_handler);
+		rfree(sad);
```

Comment:
rfree(NULL) is a no-op so you don't have to check for nullity.

Reply:
`if (sad)` was intended to guard `sad->model_handler`. Currently all jumps to `fail` has `sad != NULL` so yes this check is unnecessary.

Reply:
Yep, that would be bad, but that is where the second label comes in like above to flatten out the code.